File: 0340-longest-substring-with-at-most-k-distinct-characters/0340-longest-substring-with-at-most-k-distinct-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        map<char,int> hashMap; 
        int answer = INT_MIN, start = 0, end = 0; 

        while(end < s.length()){
            hashMap[s[end]]++; 

            while(hashMap.size() > k){
                hashMap[s[start]]--; 
                if(hashMap[s[start]] == 0) hashMap.erase(s[start]); 
                start++; 
            }

            if(hashMap.size() <= k){
                answer = max(answer, end - start + 1); 
            }
            end++; 
        }


        return answer; 
    }
};

File: 0340-longest-substring-with-at-most-k-distinct-characters/0340-longest-substring-with-at-most-k-distinct-characters.java
class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        int answer = 0; 
        int start = 0, end = 0; 
        Map<Character,Integer> hashMap = new HashMap<>(); 
        while(end < s.length()){
            char curr = s.charAt(end); 
            hashMap.merge(curr,1,Integer::sum); 
            while(hashMap.size() > k){
                char prev = s.charAt(start);
                hashMap.merge(prev,-1,Integer::sum); 
                if(hashMap.get(prev) <= 0) hashMap.remove(prev); 
                start++; 
            }
            answer = Math.max(answer, end - start + 1); 
            end++; 
        }

        return answer; 
    }
}